astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -627,6 +655,51 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
                     pass
 
 
+@dataclasses.dataclass
+class RestoreUserDefinedFunctionsStep(Step[None]):
+    zookeeper_client: ZooKeeperClient
+    replicated_user_defined_zookeeper_path: str | None
+    clients: Sequence[ClickHouseClient]
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
+        if self.replicated_user_defined_zookeeper_path is None:
+            return
+
+        clickhouse_manifest = context.get_result(ClickHouseManifestStep)
+        if not clickhouse_manifest.user_defined_functions:
+            return
+
+        async with self.zookeeper_client.connect() as connection:
+            for user_defined_function in clickhouse_manifest.user_defined_functions:
+                path = os.path.join(self.replicated_user_defined_zookeeper_path, user_defined_function.path)
+                await connection.create(path, user_defined_function.create_query)
+
+        # wait for all user defined functions to be restored from zk
+        create_queries: list[str] = []
+        for function in clickhouse_manifest.user_defined_functions:
+            create_queries.append(escape_sql_string(function.create_query.removesuffix(b"\n")))
+        create_queries_str = ", ".join(create_queries)
+        query = f"""\
+            SELECT count(*)
+            FROM system.functions
+            WHERE create_query IN ({create_queries_str})
+            AND origin = 'SQLUserDefined'
+        """.encode()
+
+        async def wait_for_functions(client: ClickHouseClient) -> None:
+            for _ in range(10):
```

Comment:
All other step timeouts are configurable, this one should be configurable too (that has been helpful during incidents).

I think you can bump it to a higher value too, like 60sec, it would be annoying to fail a restore just because there are many functions or ZK/ClickHouse struggled temporarily for ~10sec.

Reply:
Can do